yjit/src/codegen.rs

```diff
@@ -1292,39 +1292,63 @@ fn gen_newrange(
 }
 
 fn guard_object_is_heap(
+    ctx: &mut Context,
     asm: &mut Assembler,
-    object_opnd: Opnd,
+    object: Opnd,
+    object_opnd: YARVOpnd,
     side_exit: Target,
 ) {
+    let object_type = ctx.get_opnd_type(object_opnd);
+    if object_type.is_heap() {
+        return;
+    }
+
     asm.comment("guard object is heap");
 
     // Test that the object is not an immediate
-    asm.test(object_opnd, (RUBY_IMMEDIATE_MASK as u64).into());
+    asm.test(object, (RUBY_IMMEDIATE_MASK as u64).into());
     asm.jnz(side_exit);
 
     // Test that the object is not false
-    asm.cmp(object_opnd, Qfalse.into());
+    asm.cmp(object, Qfalse.into());
     asm.je(side_exit);
+
+    if object_type.diff(Type::UnknownHeap) != usize::MAX {
+        ctx.upgrade_opnd_type(object_opnd, Type::UnknownHeap);
+    }
```

Comment:
This PR is a good idea. Good work Kokubun 👍 

One thing that I think we may want to change is that this pattern is pretty common IMO, where we do a test before calling `ctx.upgrade_opnd_type`. I think that probably, we should have some logic inside `upgrade_opnd_type` that checks if the "upgrade" gives us more info than we already had, and that we're not contradicting type info we already had. Then we wouldn't need to have if statements everywhere, we can just call `upgrade_opnd_type` with the type information we know, and the context object takes care of the rest. Doesn't have to be done as part of this PR though, just an idea for another improvement.

Reply:
> I think that probably, we should have some logic inside upgrade_opnd_type that checks if the "upgrade" gives us more info than we already had, and that we're not contradicting type info we already had.

We already assert that with `Type#upgrade`, and the problem here is that we would trip the assertion without this `if` statement. That case should probably be `CantCompile` as Alan said, but it seems like that would make `ratio_in_yjit` worse, so I'm not changing the codegen strategy in this PR yet. I plan to address that in a separate PR.